File: pda_scanner/lib/src/pda_listener.dart
import 'package:flet/flet.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:my_pda_scanner/my_pda_scanner_mixin.dart';
import 'package:my_pda_scanner/my_pda_scanner_util.dart';

class PdaListener extends StatefulWidget {
  final Control? parent;
  final Control control;
  final FletControlBackend backend;

  const PdaListener(
      {super.key, this.parent, required this.control, required this.backend});
  @override
  _PdaListener createState() => _PdaListener();
}

class _PdaListener extends State<PdaListener>
    with MyPdaScannerMixin<PdaListener> {
  String _pda_code = "";
  bool _start_listener = false;
  // int _event_count = 0;
  late TextEditingController _controller;
  MyPdaScannerUtil pdaScannerUtil = MyPdaScannerUtil();

  // MyPdaScannerUtil pdaScannerUtil = MyPdaScannerUtil();

  @override
  Widget build(BuildContext context) {
    var pda_code = widget.control.attrString("pda_code");
    _controller = TextEditingController();
    // if (_pda_code != pda_code) {
    //   pda_code = _pda_code;
    //   _controller.text = _pda_code;
    // }
    if (_pda_code != "") {
      pda_code = _pda_code;
      _controller.text = _pda_code;
      // widget.backend
      // .updateControlState(widget.control.id, {"pda_code": pda_code});
    }
    var pda_action = widget.control.attrString("pda_action");
    var data_tag = widget.control.attrString("data_tag");
    var hint_text = widget.control.attrString("hint_text");
    // 测试时为观察事件数量创建的变量
    // var event_count = widget.control.attrInt("event_count");

    bool onChange = widget.control.attrBool("onChange", false)!;
    bool start_listener = widget.control.attrBool("start_listener")!;
    pdaScannerUtil.sendMessageToAndroid(pda_action!, data_tag!);
    final ValueChanged<String>? _on_changed;
    if (start_listener) {
      setState(() {
        _start_listener = true;
      });
    } else {
      setState(() {
        _start_listener = false;
      });
    }

    Widget pda_control = TextField(
        controller: _controller,
        decoration: InputDecoration(
          border: InputBorder.none, //无边框
          hintText: hint_text, //提示文本
        ),
        onChanged: (String value) {
          debugPrint(value);
          _pda_code = value;
          widget.backend
              .updateControlState(widget.control.id, {"pda_code": value});
          if (onChange) {
            widget.backend
                .triggerControlEvent(widget.control.id, "change", value);
          }
        });

    return constrainedControl(
        context, pda_control, widget.parent, widget.control);
  }

  @override
  Future<void> myPdaScannerCodeHandle(String code) async {
    // 编写你的逻辑
    if (_start_listener) {
      setState(() {
        _pda_code = code;
      });
      widget.backend.updateControlState(widget.control.id, {"pda_code": code});
      widget.backend.triggerControlEvent(widget.control.id, "listener", code);
    }
  }
}

File: python/controls/pda_listener.py
from flet_core.constrained_control import ConstrainedControl
from typing import Optional


class PdaListener(ConstrainedControl):
    """
    PdaListener 控件。
    """

    def __init__(
            self,
            pda_code: Optional[str] = None,
            pda_action: Optional[str] = None,
            data_tag: Optional[str] = None,
            start_listener: Optional[bool] = False,    # 初始为不监听广播
            hint_text: Optional[str] = None,
            on_change = None,
            on_listener = None,
        ):
        ConstrainedControl.__init__(self)
        self.pda_code = pda_code
        self.pda_action = pda_action
        self.data_tag = data_tag
        self.start_listener = start_listener
        self.hint_text = hint_text
        self.on_change = on_change
        self.on_listener = on_listener



    def _get_control_name(self):
        return "pda_listener"

    @property
    def pda_action(self):
        return self._get_attr("pda_action")

    @pda_action.setter
    def pda_action(self, value):
        self._set_attr("pda_action", value)

    @property
    def data_tag(self):
        return self._get_attr("data_tag")

    @data_tag.setter
    def data_tag(self, value):
        self._set_attr("data_tag", value)

    @property
    def pda_code(self) -> Optional[str]:
        return self._get_attr("pda_code", def_value="")

    @pda_code.setter
    def pda_code(self, value):
        self._set_attr("pda_code", value)

    @property
    def start_listener(self):
        return self._get_attr("start_listener", def_value=False)

    @start_listener.setter
    def start_listener(self, value):
        self._set_attr("start_listener", value)

    @property
    def hint_text(self):
        return self._get_attr("hint_text")

    @hint_text.setter
    def hint_text(self, value):
        self._set_attr("hint_text", value)

    # on_change
    @property
    def on_change(self):
        return self._get_event_handler("change")

    @on_change.setter
    def on_change(self, handler):
        self._add_event_handler("change", handler)
        self._set_attr("onChange", True if handler is not None else None)


    @property
    def on_listener(self):
        return self._get_event_handler("listener")

    @on_listener.setter
    def on_listener(self, handler):
        self._add_event_handler("listener", handler)
        # self._set_attr("start_listener", True if handler is not None else None)

     